File: npc_sim/simulation/population_stats.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Per-tick aggregate population statistics."""

from __future__ import annotations
from collections import deque
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from npc_sim.npc.npc import NPC
    from npc_sim.events.sim_event import SimEvent


class PopulationStats:
    """Real-time aggregate metrics for the entire population."""

    def __init__(self):
        self.total_population: int = 0
        self.avg_hunger: float = 0.0
        self.avg_thirst: float = 0.0
        self.avg_happiness: float = 0.0
        self.avg_stress: float = 0.0
        self.total_deaths: int = 0
        self.recent_combat_count: int = 0
        self.is_at_war: bool = False

    def update(self, npcs: list, event_log) -> None:
        alive = [n for n in npcs if n.is_active and n.vitals.is_alive]
        self.total_population = len(alive)

        if not alive:
            return

        self.avg_hunger = sum(n.vitals.hunger for n in alive) / len(alive)
        self.avg_thirst = sum(n.vitals.thirst for n in alive) / len(alive)
        self.avg_happiness = sum(n.psychology.happiness for n in alive) / len(alive)
        self.avg_stress = sum(n.vitals.stress for n in alive) / len(alive)

        # Count recent combat events
        self.recent_combat_count = sum(
            1 for e in event_log if e.category == "combat"
        )
        self.is_at_war = self.recent_combat_count > 5

    def is_famine(self) -> bool:
        return self.avg_hunger > 0.7

    def is_prosperous(self) -> bool:
        return self.avg_happiness > 0.5 and self.avg_hunger < 0.3

    def to_dict(self) -> dict:
        return {
            "total_population": self.total_population,
            "avg_hunger": round(self.avg_hunger, 3),
            "avg_thirst": round(self.avg_thirst, 3),
            "avg_happiness": round(self.avg_happiness, 3),
            "avg_stress": round(self.avg_stress, 3),
            "is_at_war": self.is_at_war,
            "is_famine": self.is_famine(),
            "is_prosperous": self.is_prosperous(),
        }

    def __repr__(self) -> str:
        return (f"[PopStats] Pop:{self.total_population} "
                f"Hunger:{self.avg_hunger:.0%} Stress:{self.avg_stress:.0%}")

File: Runtime/Simulation/PopulationStats.cs
using System;
using System.Collections.Generic;

namespace ForgeProject.Sim
{
    /// <summary>
    /// Aggregate civilisation-level statistics tracked across all NPCs.
    /// Updated once per sim-tick by <see cref="SimulationManager"/>.
    /// Useful for dashboards, scenario victory conditions, and AI analytics.
    /// </summary>
    public sealed class PopulationStats
    {
        // ─── Population ───────────────────────────────────────────────────────────
        public int   TotalNPCs        { get; private set; }
        public int   AliveNPCs        { get; private set; }
        public int   DeadThisTick     { get; private set; }

        // ─── Averages ─────────────────────────────────────────────────────────────
        public float AvgHealth        { get; private set; }
        public float AvgHunger        { get; private set; }
        public float AvgThirst        { get; private set; }
        public float AvgEnergy        { get; private set; }
        public float AvgStress        { get; private set; }
        public float AvgHappiness     { get; private set; }
        public float AvgFear          { get; private set; }
        public float AvgReputation    { get; private set; }

        // ─── Events this tick ─────────────────────────────────────────────────────
        public int   CombatEventsThisTick  { get; private set; }
        public int   TradeEventsThisTick   { get; private set; }
        public int   SocialEventsThisTick  { get; private set; }

        // ─── Historical ring ──────────────────────────────────────────────────────
        private const int HistoryLength = 100;
        private readonly Queue<float> _happinessHistory = new(HistoryLength);
        private readonly Queue<float> _hungerHistory    = new(HistoryLength);

        public IReadOnlyCollection<float> HappinessHistory => _happinessHistory;
        public IReadOnlyCollection<float> HungerHistory    => _hungerHistory;

        // ─── Update ───────────────────────────────────────────────────────────────

        /// <summary>
        /// Recomputes all statistics from the live NPC list.
        /// Called once per tick by <see cref="SimulationManager"/>.
        /// </summary>
        public void Update(IReadOnlyList<NPC> npcs, IReadOnlyCollection<SimEvent> events)
        {
            TotalNPCs     = npcs.Count;
            DeadThisTick  = 0;
            int alive     = 0;
            float h = 0, hunger = 0, thirst = 0, energy = 0, stress = 0, happy = 0, fear = 0, rep = 0;

            foreach (var npc in npcs)
            {
                if (!npc.Vitals.IsAlive) { DeadThisTick++; continue; }
                alive++;
                h      += npc.Vitals.Health;
                hunger += npc.Vitals.Hunger;
                thirst += npc.Vitals.Thirst;
                energy += npc.Vitals.Energy;
                stress += npc.Vitals.Stress;
                happy  += npc.Psychology.Happiness;
                fear   += npc.Psychology.Fear;
                rep    += npc.Social.Reputation;
            }

            AliveNPCs    = alive;
            float inv    = alive > 0 ? 1f / alive : 0f;
            AvgHealth    = h      * inv;
            AvgHunger    = hunger * inv;
            AvgThirst    = thirst * inv;
            AvgEnergy    = energy * inv;
            AvgStress    = stress * inv;
            AvgHappiness = happy  * inv;
            AvgFear      = fear   * inv;
            AvgReputation= rep    * inv;

            // Event breakdown
            CombatEventsThisTick = SocialEventsThisTick = TradeEventsThisTick = 0;
            foreach (var ev in events)
            {
                switch (ev.Category)
                {
                    case "combat":   CombatEventsThisTick++;  break;
                    case "social":   SocialEventsThisTick++;  break;
                    case "economy":  TradeEventsThisTick++;   break;
                }
            }

            // History ring
            if (_happinessHistory.Count >= HistoryLength) _happinessHistory.Dequeue();
            _happinessHistory.Enqueue(AvgHappiness);
            if (_hungerHistory.Count >= HistoryLength) _hungerHistory.Dequeue();
            _hungerHistory.Enqueue(AvgHunger);
        }

        // ─── Conditions ───────────────────────────────────────────────────────────

        /// <summary>Returns true when average happiness is above the given threshold (default: 0.5).</summary>
        public bool IsProsperous(float threshold = 0.5f) => AvgHappiness > threshold && AvgHunger < 0.4f;

        /// <summary>Returns true when average hunger exceeds the given threshold — a famine condition.</summary>
        public bool IsFamine(float threshold = 0.7f) => AvgHunger > threshold;

        /// <summary>Returns true when combat events dominate social events (war-like state).</summary>
        public bool IsAtWar => CombatEventsThisTick > SocialEventsThisTick * 2;

        public override string ToString() =>
            $"[PopStats] Pop:{AliveNPCs} AvgHappy:{AvgHappiness:+0.00;-0.00} AvgHunger:{AvgHunger:P0} Combat:{CombatEventsThisTick} Trade:{TradeEventsThisTick}";
    }
}
